### ISSUE TITLE
Type of class member functions does not contain this type

### ISSUE THREAD

**Issue by thebanjomatic** [microsoft/TypeScript#36803]
It appears as though the `this` type of a class member function is not being reported as part of the type definition for the function and instead it is inferred as `unknown` even though usage of `this` within the function behaves appropriately.

**TypeScript Version:** 
3.7.5
```
noImplicitThis
strictFunctionTypes
```

**Search Terms:**
Member Function This Type
Extract This Type

**Code**

```ts
class Foo {
    value: number = 0;
    func1() {
        this.value++;
    }

    func2(this: Foo) {
        this.value--;
    }

    func3 = function() {
        // can't access this.value
    }
}

type FuncType1 = Foo['func1'];
type FuncType2 = Foo['func2'];
type FuncType3 = Foo['func3'];

type InferThisType<T> = T extends (this: infer THIS, ...args: any) => any ? THIS : unknown;

type This1 = InferThisType<Foo['func1']>;
type This2 = InferThisType<Foo['func2']>;
type This3 = InferThisType<Foo['func3']>;
```

**Expected behavior:**
```ts
type FuncType1 = Foo['func1']; // (this: Foo) => void
type FuncType2 = Foo['func2']; // (this: Foo) => void
type FuncType3 = Foo['func3']; // () => void

type This1 = InferThisType<Foo['func1']>; // Foo
type This2 = InferThisType<Foo['func2']>; // Foo
type This3 = InferThisType<Foo['func3']>; // unknown
```

**Actual behavior:**
```ts
type FuncType1 = Foo['func1']; // () => void
type FuncType2 = Foo['func2']; // (this: Foo) => void
type FuncType3 = Foo['func3']; // () => void

type This1 = InferThisType<Foo['func1']>; // unknown
type This2 = InferThisType<Foo['func2']>; // Foo
type This3 = InferThisType<Foo['func3']>; // unknown
```

**Playground Link:**
https://www.typescriptlang.org/play/?ssl=20&ssc=120&pln=20&pc=1#code/MYGwhgzhAEBiD29oG8BQ0PQG5hAVwFMAuaAOzwFsAjAgJ2gF5oAGAbnUwDM9TgBGABQBKFB0yYALgAsAlhAB0OfAQDUK9uIC+qMdG68ATAOlySCeCLTjxJhUsIBaBxszbd+4AGZGensAky8KTCotaYAPTh0MBgpADkEtBgwMAEUNC2iriEutpuEgCeAA4EcH4AKsUEfD7mANpxHnxxALqs0JHQIQwAfNjwMgAmqIUlZbyVJQa1iA0eBq3tncayEGaIIr39QyNV48CTBN5M9Y1+nosdUd19WAPDu2MAkqScdOWrhwA85X1M5dACAAPCQEUiDGArUzQGSvOjQcoACSeAGUADTQeRYsC0ADmaySpAKmz6sQK0AA-AjkSjoCQeABrUjwADupHYj1KHzkNSYLzetG5EG+pyarR6SyijOZbM5CNW0z5cMFnyqX1FfgWLQlVzgiDlQuO0H571VJXVszOvAu2sl0GlrNIqCAA

---

**Motivation**
I am trying to create a mechanism for creating a type which excludes member-function properties, but my plan was foiled due to not being able to infer the `this` type for the function.

```ts
type InferThisType<T> = T extends (this: infer THIS, ...args: any) => any ? THIS : unknown;
type IsMemberFunction<T, P extends keyof T> = InferThisType<T[P]> extends T ? true : false;

type NonMemberFunctionPropertyKeys<T> = {
   [P in keyof T]: IsMemberFunction<T, P> extends true ? never : P
}[keyof T];

type ExcludeMemberFunctions<T> = Pick<T, NonMemberFunctionPropertyKeys<T>>;
```

I'd like to use this to provide stronger typing guarantees for angular bindings as well as the benefits of code-completion, etc.
```ts
type AngularBindings<T extends angular.IComponentController> = 
  Partial<
    Record<
      Exclude<NonMemberFunctionPropertyKeys<T>, keyof angular.IComponentController>,
      '@' | '<' | '=' | '&' | '@?' | '<?' | '=?' | '&?'
    >
  >;
```
```ts
class MyController implements angular.IComponentController {
  public value: number;
  public callback: Function;

  public memberFunc() {
    this.value++;
    this.callback();
  }
}

class MyComponent implements angular.IComponentOptions {
  public bindings: AngularBindings<MyController>;
  public controller: angular.Injectable<angular.IControllerConstructor>;
  public template: string;

   constructor() {
     this.bindings = {
       value: '<',
       callback: '<',
       // memberFunc would produce compile error as would any typo.
     };
     this.template = html;
     this.controller = MyController;
  }
}
```

**Related Issues:**

**Comment by thebanjomatic** [comment]
Additionally, it would seem that fixing this could catch a class of errors when grabbing functions off of a class instance and calling them without `call` or `apply`
```ts
const f = new Foo();
const f1 = f.func1;
const f2 = f.func2;

f1(); // Should produce an error as calling f1() will throw.
f2(); // Error: The 'this' context of type 'void' is not assignable to method's 'this' of type 'Foo'. (2684)
```

https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=17&pc=6#code/MYGwhgzhAEBiD29oG8BQ0PQG5hAVwFMAuaAOzwFsAjAgJ2gF5oAGAbnUwDM9TgBGABQBKFB0yYALgAsAlhAB0OfAQDUK9uIC+qMdG68ATAOlySCeCLTjxJhUsIBaBxszbtqYPFIQJexmQIAdzhEYXZPb19OPn9OeX1+cK8fPQNY+J5gA3ZUaLDcoyFWIA

**Comment by RyanCavanaugh** [comment]
We don't implicitly add `this` to methods because it ends up making every function generic, which caused a 20%+ performance hit for compilation times in class-heavy codebases. You can "opt in" to methods which are in danger of being pulled off.
